### ISSUE TITLE
triqs::atom_diag::quantum_number_eigenvalues returns incorrect values when constraining the number of particles

### ISSUE THREAD

**Issue by YannIntVeld** [TRIQS/triqs#825]
### Description

We have been using the atom_diag class in Python to diagonalize small Hubbard systems. As we are only interested in fixed occupation numbers, we constrained the diagonalization to certain particle number subspaces. (By modifying the installation slightly as proposed by @the-hampel ) However, we find that the 'quantum_number_eigenvalues' functions no longer return correct information.

### Steps to Reproduce

First we modified the TRIQS installation slightly in order to be able to constrain atom_diag to fixed particle numbers in Python.
In https://github.com/TRIQS/triqs/blob/unstable/python/triqs/atom_diag/atom_diag_desc.py we added a new method call wrapper in line 64
```c++
c.add_constructor("(many_body_operator h, fundamental_operator_set fops, int n_min, int n_max)", doc = "Reduce a given Hamiltonian to a block diagonal, diagonalize only within particle numbers given")
```
and recompiled TRIQS.

Next we used a simple two-site tight-binding model to test our results. Our minimal code example:
```python
from triqs.operators import *
from triqs.operators.util.observables import *
from triqs.atom_diag import *
import numpy as np
from itertools import product

def get_multiplet_occ(atom_diag, orb_names, spin_names = ['up','down'], off_diag = True):
    '''
    function to extract operator expectation values from measured 
    density matrix and atom_diag object from cthyb solver.
    
    measures N, and constructs eigenstates, and probabilities
    
    Parameters
    ----------
    atom_diag: triqs atom_diag object
        contains information about the local Hamiltonian (Hloc_0 + H_int)
    orb_names : list of int
        list of orbital indices
    spin_names : list of string
        list of strings containing the spin channel names
    off_diag: boolean
        determines wether blocks of Gf are named up_0 (false) or just up (true)
    __Returns:__
    res : Panda DataFrame
        containing all results structured
    '''
    import pandas as pd
    
    
    # res will be a list of dictionaries to be a panda data frame
    res = []
    
    # get fundamental operators from atom_diag object
    occ_operators = []
    for op in atom_diag.fops:
        occ_operators.append( n(op[0], op[1]) )
    
    # construct total occupation operator from list
    N_op = sum(occ_operators)
        
    # get particle numbers from atom_diag
    particle_numbers = quantum_number_eigenvalues(N_op, atom_diag)
    print(particle_numbers)
    N_max = int(max(map(max, particle_numbers)))

    for sub in range(0,atom_diag.n_subspaces):   
        
        # first get fock space spanning the subspace
        fs_states = []
        for ind, fs in enumerate(atom_diag.fock_states[sub]):
            # get state in binary representation
            state =int(bin(int(fs))[2:])
            fmt='{0:0'+str(N_max)+'d}'
            fs_states.append("|"+fmt.format(state)+">")
            
        # now extract particle number etc.
        for ind in range(atom_diag.get_subspace_dim(sub)):
            
            # get particle number
            # carefully here to not cast to int as the particle number
            # can be something like 1.999999996 and would get then 1! 
            particle_number = round(particle_numbers[sub][ind])
            if abs(particle_number-particle_numbers[sub][ind]) > 1e-12:
                raise ValueError('round error for particle number to large!',particle_numbers[sub][ind])
            else:
                particle_number = int(particle_number)
            # energy of state
            eng=atom_diag.energies[sub][ind]
            
            # construct eigenvector in fock state basis:
            ev_state = ''
            for i, elem in enumerate(atom_diag.unitary_matrices[sub][:,ind]):
                ev_state += ' {:+1.4f}'.format(elem)+fs_states[i]
            
            
            # add to data frame
            res.append({"Sub#" : sub, 
                        "EV#" : ind, 
                        "N" : particle_number, 
                        "energy" : eng,
                        "state": ev_state})
    # panda data frame from res
    res = pd.DataFrame(res, columns=res[0].keys())
    
    return res


# Define operators
spin_names = ['up', 'down']
orb_names = ['0', '1']

fops = [ (sn, on) for sn, on in product(spin_names, orb_names) ]

# Setup hamiltonian
t = 1.0
HH = -t * ( c_dag('up', '0') * c('up', '1') + c_dag('down', '0') * c('down', '1'))\
     -t * ( c_dag('up', '1') * c('up', '0') + c_dag('down', '1') * c('down', '0'))

# Diagonalize
ad = AtomDiag(HH, fops, 0, 1)
adOrig = AtomDiag(HH, fops)

res2 = get_multiplet_occ(ad, orb_names=['0','1'], spin_names = ['up','down'], off_diag = True)
print(res2.sort_values('N', ascending=True))

print('--------------------')

res = get_multiplet_occ(adOrig, orb_names=['0','1'], spin_names = ['up','down'], off_diag = True)
print(res.sort_values('N', ascending=True))

```

Here `ad` is the solved diagonalization problem when constraining the occupation number to 0 and 1, and `adOrig` is the solved diagonalization problem without constraining the occupation number.

**Expected behavior:** We expect the particle number eigenvalues of `ad` to contain one zero-particle state and four one-particle states.

**Actual behavior:** All particle number eigenvalues of `ad` return as 0.

The raw output is:
```
[array([0., 0.]), array([0., 0.]), array([0.])]
   Sub#  EV#  N  energy                        state
0     0    0  0     0.0       -0.7071|1> -0.7071|10>
1     0    1  0     2.0       -0.7071|1> +0.7071|10>
2     1    0  0     0.0   -0.7071|100> -0.7071|1000>
3     1    1  0     2.0   -0.7071|100> +0.7071|1000>
4     2    0  0     1.0                   +1.0000|0>
--------------------
[array([2., 2., 2., 2.]), array([1., 1.]), array([1., 1.]), array([3., 3.]), array([3., 3.]), array([0.]), array([2.]), array([2.]), array([4.])]
    Sub#  EV#  N  energy                                              state
12     5    0  0     2.0                                       +1.0000|000>
4      1    0  1     1.0                          -0.7071|001> -0.7071|010>
5      1    1  1     3.0                          -0.7071|001> +0.7071|010>
6      2    0  1     1.0                         -0.7071|100> -0.7071|1000>
7      2    1  1     3.0                         -0.7071|100> +0.7071|1000>
0      0    0  2     0.0   -0.5000|101> -0.5000|110> -0.5000|1001> -0.50...
1      0    1  2     2.0   -0.7071|101> -0.0000|110> -0.0000|1001> +0.70...
2      0    2  2     2.0   +0.0000|101> -0.7071|110> +0.7071|1001> +0.00...
3      0    3  2     4.0   +0.5000|101> -0.5000|110> -0.5000|1001> +0.50...
13     6    0  2     2.0                                       +1.0000|011>
14     7    0  2     2.0                                      +1.0000|1100>
8      3    0  3     1.0                         -0.7071|111> -0.7071|1011>
9      3    1  3     3.0                         -0.7071|111> +0.7071|1011>
10     4    0  3     1.0                        -0.7071|1101> -0.7071|1110>
11     4    1  3     3.0                        -0.7071|1101> +0.7071|1110>
15     8    0  4     2.0                                      +1.0000|1111>
```

We found similarly incorrectly returned values for eigenvalues of other operators, such as the total energy operator or the total spin operator.

**Comment by Wentzell** [comment]
Dear @YannIntVeld,

Thank you for pointing out this bug and for providing the example.
We will look into this and include a fix for the coming 3.1 release.

@krivenko Do you have any idea what may be going wrong here?

**Comment by krivenko** [comment]
@Wentzell Should be resolved by #831.

**Comment by the-hampel** [comment]
Hi @krivenko,
I just compiled your PR and can confirm that the problem is resolved. The output for the test above is now correct and shows:
```
[array([1., 1.]), array([1., 1.]), array([0.])]
   Sub#  EV#  N  energy                        state
4     2    0  0     1.0                   +1.0000|0>
0     0    0  1     0.0       -0.7071|1> -0.7071|10>
1     0    1  1     2.0       -0.7071|1> +0.7071|10>
2     1    0  1     0.0   -0.7071|100> -0.7071|1000>
3     1    1  1     2.0   -0.7071|100> +0.7071|1000>
```
which is correct. 

Thank you for the quick fix! 

Best,
Alex

**Comment by Wentzell** [comment]
Thank you @krivenko for quickly providing a fix!

**Comment by YannIntVeld** [comment]
Dear all,

Thanks a lot for the quick fix, this is already very helpful!
However unfortunately I believe the problem is not yet completely fixed.

When I use `ad = AtomDiag(HH, fops, 0, 1)` the outout is indeed correct.
However, when I use `ad = AtomDiag(HH, fops, 1, 2)` I get the following output:

```[array([2., 2., 2., 2.]), array([0., 0.]), array([0., 0.]), array([2.]), array([2.])]
   Sub#  EV#  N  energy                                              state
4     1    0  0     1.0                             -0.7071|1> -0.7071|10>
5     1    1  0     3.0                             -0.7071|1> +0.7071|10>
6     2    0  0     1.0                         -0.7071|100> -0.7071|1000>
7     2    1  0     3.0                         -0.7071|100> +0.7071|1000>
0     0    0  2     0.0   -0.5000|101> -0.5000|110> -0.5000|1001> -0.50...
1     0    1  2     2.0   -0.7071|101> -0.0000|110> -0.0000|1001> +0.70...
2     0    2  2     2.0   +0.0000|101> -0.7071|110> +0.7071|1001> +0.00...
3     0    3  2     4.0   +0.5000|101> -0.5000|110> -0.5000|1001> +0.50...
8     3    0  2     2.0                                        +1.0000|11>
9     4    0  2     2.0                                      +1.0000|1100>
```

I.e., I find four zero-particle states, and six two-particle states. I would expect to find four one-particle states instead. 
I believe any eigenvalue calculated with `quantum_number_eigenvalues()` is incorrect when n_min > 0.

Would you like me to create a new issue for this?

**Comment by Wentzell** [comment]
The issue looks to be directly related, so I will reopen this.
@krivenko Could I ask you again to have a look at this?

**Comment by krivenko** [comment]
Okay, now this is nasty. I have to say that the issue reported in https://github.com/TRIQS/triqs/issues/825#issuecomment-1032758218 is a conceptual one.

`quantum_number_eigenvalues()` as well as many other functions from `atom_diag` uses matrix elements and subspace-to-subspace connections of annihilation/creation operators, which are computed upon `atom_diag`'s construction.

Let's say we restrict the total number of particles to the range [1; 2]. This results in a partial Hilbert space constructed as a direct sum of 5 subspaces. As it is seen from a script's output above, subspace 1 should contain one particle. State from which subspace would we get if we acted with an annihilation operator on a state from subspace 1? 
```python
print("# subspaces:", ad.n_subspaces)
print(f"c_{fops[0]} |1> ->", ad.c_connection(0, 1))
print(f"c_{fops[1]} |1> ->", ad.c_connection(1, 1))
print(f"c_{fops[2]} |1> ->", ad.c_connection(2, 1))
print(f"c_{fops[3]} |1> ->", ad.c_connection(3, 1))
```
```
# subspaces: 5
c_('up', '0') |1> -> -1
c_('up', '1') |1> -> -1
c_('down', '0') |1> -> -1
c_('down', '1') |1> -> -1
```
In other words, none of the annihilation operators generates a valid connection because *there is nothing to connect to!*
The subspace containing the vacuum state has already been excluded from consideration.

Since occupation number operators are products of `c_dag` and `c`, they do not generate valid connections from subspace 1 either. This explains why one gets zeros for all occupations in subspace 1.

**Comment by krivenko** [comment]
@Wentzell @the-hampel Do you have any ideas on how to proceed here? 

I am afraid that making `quantum_number_eigenvalues()` work as expected with partial Hilbert spaces would require a complete redesign of `atom_diag`.

**Comment by Wentzell** [comment]
Dear @krivenko,

Thank you for looking into this! I understand the conceptual problem.

In this example we really only care about the fact that `n` takes you from a valid state,
i.e. a state that is a sum of fock-states fulfilling our particle number constraint, to another valid state.
In other words we really want to permit, between the fundamental operators `c_dag` and `c_d`,
intermediate states that do not fulfill the particle number constraint, as long as the final state 
after applying the monomial is valid. Do you agree?

I looked at the current implementation and I think this could be achieved without too much effort in fact.
The function that forms the matrix for each monomial is
https://github.com/TRIQS/triqs/blob/617a40c737688805fc553ca70b64b337d025c648/c%2B%2B/triqs/atom_diag/impl/atom_diag.cpp#L102-L121

Currently it does so by using `cdag_matrix` and `c_matrix` which are represented in the eigenbasis.
Couldn't we instead
- Calculate the monomial matrix in the fock state basis
- Allow any intermediate fockstate
- Check that the final fockstate is valid
- Return the matrix after rotating it to the eigenbasis

This way we would be completely fine with the fact that we have diagonalized the Hamiltonian only in
our constrained subspace.

**Comment by krivenko** [comment]
Dear @Wentzell,

Thank you for the input!
I guess you are right, we could sacrifice some speed and build monomial matrices from scratch using `imperative_operator`s acting in the full Hilbert space (unless I am missing something...)

Let me try drafting a solution that we can discuss in more detail.

**Comment by krivenko** [comment]
Please, check that  #835 properly solves the issue.

**Comment by Wentzell** [comment]
I can confirm that this issue is now fixed after the merge of [3e16235](https://github.com/TRIQS/triqs/commit/3e1623508821d3f926c3cb73695dabd98599b35e). Thank you @krivenko for looking into this!

**Comment by Wentzell** [comment]
@YannIntVeld I noticed that the output that you posted hear deviates from I am getting for your example.
This already applies for the full diagonalization without particle number constraints.
I am getting 

```
    Sub#  EV#  N  energy                                              state
12     5    0  0     2.0                                      +1.0000|0000>
4      1    0  1     1.0                        -0.7071|0001> -0.7071|0010>
5      1    1  1     3.0                        -0.7071|0001> +0.7071|0010>
6      2    0  1     1.0                        -0.7071|0100> -0.7071|1000>
7      2    1  1     3.0                        -0.7071|0100> +0.7071|1000>
0      0    0  2     0.0   +0.5000|0101> +0.5000|0110> +0.5000|1001> +0....
1      0    1  2     2.0   +0.5000|0101> -0.5000|0110> +0.5000|1001> -0....
2      0    2  2     2.0   -0.5000|0101> -0.5000|0110> +0.5000|1001> +0....
3      0    3  2     4.0   +0.5000|0101> -0.5000|0110> -0.5000|1001> +0....
13     6    0  2     2.0                                      +1.0000|0011>
14     7    0  2     2.0                                      +1.0000|1100>
8      3    0  3     1.0                        -0.7071|0111> -0.7071|1011>
9      3    1  3     3.0                        -0.7071|0111> +0.7071|1011>
10     4    0  3     1.0                        -0.7071|1101> -0.7071|1110>
11     4    1  3     3.0                        -0.7071|1101> +0.7071|1110>
15     8    0  4     2.0                                      +1.0000|1111>
```
while your output shows the following states in the two-particle sector

```
1      0    1  2     2.0   -0.7071|101> -0.0000|110> -0.0000|1001> +0.70...
2      0    2  2     2.0   +0.0000|101> -0.7071|110> +0.7071|1001> +0.00...
```

Did you run the example exactly as posted above, or did you make alterations?

**Comment by YannIntVeld** [comment]
Dear all,
Thank you very much for the quick fix. The issue when constraining the particle number has also been fixed on my side.

@Wentzell This is very well spotted. I ran the example exactly as posted above.
I tried to run the code in both the unstable branch and the stable 3.0.x branch, using exactly the example posted above without particle number constraints.
In both cases I get the same result as I originally posted.

**Comment by Wentzell** [comment]
I think I know what is happening here. The subspace for `N=2` and `energy=2.0` is degenerate.
In other words the basis in that subspace is not uniquely defined. It seems that in this case
the eigenstates are dependent on the libraries used for the diagonalization.
